File: longest-valid-paren/one.py

class Solution(object):
        

    def longestValidParentheses(self, s):
        """
        :type s: str
        :rtype: int
        """
        def isSame(op, cp):
            if op == '(' and cp == ')':
                return True
            elif op == '[' and cp == ']':
                return True
            elif op == '{' and cp == '}':
                return True
            return False
    
        cur_longest = 0
        stk = []
        is_bal = True
        l = 0
        for c in s:
            l += 1
            if c in '([{':
                stk.append(c)
            else:
                if not stk:
                    is_bal = False
                else:
                    last_paren = stk.pop()
                    if not isSame(last_paren, c):
                        is_bal = False
        
            if len(stk) == 0 and is_bal:
                cur_longest = max(l, cur_longest)

            if not is_bal:
                l = 0
                stk.clear()
                is_bal = True
        
        return cur_longest

        
        

s = Solution()

print(s.longestValidParentheses("()(()"))
        

File: longest-valid-paren/one.cpp
#include <iostream>
#include <string>
#include <stack>
#include <algorithm> // Include this to use std::max

using namespace std;

class Solution
{
public:
    int longestValidParentheses(string s)
    {
        stack<int> stk;
        int mx = 0; // Initialize mx to 0
        int n = s.size();

        // Initialize the stack with a base index of -1
        stk.push(-1);

        for (int i = 0; i < n; i++)
        {
            if (s[i] == '(')
            {
                stk.push(i); // Push the index of '('
            }
            else
            {
                stk.pop(); // Pop the top element
                if (stk.empty())
                {
                    // If stack is empty, push the current index as a new base
                    stk.push(i);
                }
                else
                {
                    // Calculate the length of the current valid substring
                    mx = max(mx, i - stk.top());
                }
            }
        }

        return mx; // Return the maximum length of valid parentheses
    }
};

int main()
{
    Solution solution;
    string s = "(()())";
    cout << "Longest valid parentheses length: " << solution.longestValidParentheses(s) << endl;
    return 0;
}
